rfd/0003-extended-approval-workflows.md

```diff
@@ -0,0 +1,272 @@
+
+### Details
+
+Let's add missing features to Teleport and use them to compose a solution
+for scenarios described in "Why" section.
+
+#### Dynamic role requests
+
+We are going to modify the `request` part of the roles to support pattern matching.
+
+```yaml
+kind: role
+metadata:
+  name: contractor
+spec:
+  options:
+    # ...
+  allow:
+    request:
+      roles: ['^customer-.*$']
+    # ...
+  deny:
+    # ...
+```
+
+#### Cluster access labels
+
+Cluster labels, similarly to `node_labels`,
+limit the access to clusters from the root cluster.
+
+```yaml
+kind: role
+metadata:
+  name: customer-a
+spec:
+  options:
+    # ...
+  allow:
+    cluster_labels:
+      customer: 'customer-a'
+    # ...
+  deny:
+    # ...
+```
+
+To preserve backwards compatibility, if not specified, `cluster_labels` default
+value is: `'*':'*'`, granting visibility and access to all clusters.
+
+#### Access request metadata
+
+Access request gets additional fields to specify information about requests
+to individual resources:
+
+```yaml
+kind: access_request
+metadata:
+  name: request-id
+spec:
+  note: 'ticket-12345`
+  user: 'bob'
+  # Role list can be empty, in case Bob is trying to access a resource
+  # and not a specific role
+  roles: []
+  # Bob has requested access to the mongodb as root
+  resources:
+  - kind: node
+    name: mongodb
+    verb: connect
+    principal: root
+```
+
+Admins will get a new CLI and API to approve requests and modify roles:
+
+```bash
+# Bob wants to ssh into mongodb as root
+$ tctl request ls
+Token             Requestor Metadata             Note         Created At (UTC)    Status
+----------------- --------- -------------------- ------------ -------------- -------
+request-id-1      bob       ssh root@mongodb     ticket-1234  07 Nov 19 19:38 UTC PENDING
+
+# craft a role granting access just to this node
+$ tctl role create custom.yaml
+
+# approve access request and assign the custom role
+$ tctl request approve request-id --roles=custom
+```
+
+#### User interface flows
+
+Imagine Bob has the following default role:
+
+```yaml
+kind: role
+metadata:
+  name: contractor
+spec:
+  allow:
+    request:
+      roles: ['^customer-.*$']
+```
+
+Users will have several options:
+
+**Option 0: Requesting access during the login in the UI or the CLI**
+
+Teleport login screen will have a separate optional field: "Request access". Alice
+clicks "Request access" expandable section and adds a note with a ticket number.
+
+Once Alice enters a reason, Teleport will create an `access_request` with note
+provided by her in case if login is successful. The user interface shows
+a screen "Requesting access, please wait" as a separate state after the login
+and redirects back to the UI or tsh.
+
+**Option 1: Requesting access after login in the UI**
+
+Once in the system, Bob clicks on the top right nav corner and
+fills in fields request access dialog window:
+
+* Selects a role from the list
```

Comment:
will the list contain all of the roles?
what if there's thousands of them (auto-generated), or a user doesn't have permission to read all roles?